sns/parity_math.c

```diff
@@ -1444,23 +1497,44 @@ M0_INTERNAL int m0_sns_ir_init(const struct m0_parity_math *math,
 	ir->si_failed_data_nr	   = 0;
 	ir->si_alive_nr		   = block_count(ir);
 
-	M0_ALLOC_ARR(ir->si_blocks, ir->si_alive_nr);
+#if ISAL_ENCODE_ENABLED
+	/* Get encode matrix */
+	ir->si_encode_matrix = math->pmi_encode_matrix;
+
+	/* Get decode table */
+	ir->si_decode_tbls = math->pmi_decode_tbls;
+
+	/* Initialize required buffers */
+	ret = isal_ir_init(ir);
+	if (ret != 0)
+		goto fini;
+
+#endif /* ISAL_ENCODE_ENABLED */
+
+	ALLOC_ARR_INFO(ir->si_blocks, ir->si_alive_nr,
+		       "blocks", ret);
 	if (ir->si_blocks == NULL)
-		return M0_ERR(-ENOMEM);
+		goto fini;
 
 	for (i = 0; i < block_count(ir); ++i) {
 		ir->si_blocks[i].sib_idx = i;
 		ir->si_blocks[i].sib_status = M0_SI_BLOCK_ALIVE;
 		ir->si_blocks[i].sib_data_recov_mat_col = IR_INVALID_COL;
 		ret = m0_bitmap_init(&ir->si_blocks[i].sib_bitmap,
 				     block_count(ir));
-		if (ret != 0)
+		if (ret != 0){
+			ret = M0_ERR_INFO(ret, "failed to initialize bitmap for"
+			                  "ir->si_blocks[%u].sib_bitmap", i);
```

Comment:
Actually, we need to handle error here:
free the bitmaps just allocated in previous loops.

Reply:
We are freeing those bitmaps in m0_sns_ir_fini(). Because of error we are jumping to 'fini' instead of 'end'.

Reply:
The same reason as previous comment.

Reply:
OK